@MicronautTest(rebuildContext = true) no longer works as of 3.5.x - context rebuild fails after first test is run

### Expected Behavior

After each test is executed, the application context should be rebuilt and the next test should be run.

### Actual Behaviour

The first test executes successfully, the second (and any subsequent) test fails to run as micronaut throws the following exception when rebuilding the context: 

```Error instantiating bean of type [io.micronaut.http.server.netty.NettyEmbeddedServerInstance]: Bean of type [class io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment] is a manually registered singleton that was registered with the context via BeanContext.registerBean(..) and cannot be created directly
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [io.micronaut.http.server.netty.NettyEmbeddedServerInstance]: Bean of type [class io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment] is a manually registered singleton that was registered with the context via BeanContext.registerBean(..) and cannot be created directly
	at app//io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2346)
	at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2282)
	at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2210)
	at app//io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1022)
	at app//io.micronaut.context.DefaultBeanContext.createBean(DefaultBeanContext.java:997)
	at app//io.micronaut.context.BeanContext.createBean(BeanContext.java:180)
	at app//io.micronaut.http.server.netty.NettyHttpServer.lambda$fireStartupEvents$15(NettyHttpServer.java:560)
	at java.base@17.0.2/java.util.Optional.ifPresent(Optional.java:178)
	at app//io.micronaut.http.server.netty.NettyHttpServer.fireStartupEvents(NettyHttpServer.java:559)
	at app//io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:301)
	at app//io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:112)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.startEmbeddedApplication(AbstractMicronautExtension.java:397)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.beforeEach(AbstractMicronautExtension.java:293)
	at app//io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeEach(MicronautJunit5Extension.java:139)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachCallbacks$2(TestMethodTestDescriptor.java:163)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$6(TestMethodTestDescriptor.java:199)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:199)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachCallbacks(TestMethodTestDescriptor.java:162)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:129)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base@17.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base@17.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at app//org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base@17.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@17.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base@17.0.2/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@17.0.2/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.UnsupportedOperationException: Bean of type [class io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment] is a manually registered singleton that was registered with the context via BeanContext.registerBean(..) and cannot be created directly
	at io.micronaut.context.NoInjectionBeanDefinition.getConstructor(NoInjectionBeanDefinition.java:131)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanDefinition(DefaultBeanContext.java:2292)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2284)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2228)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:2989)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:2891)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2852)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2773)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1594)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:66)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2055)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1285)
	at io.micronaut.http.server.netty.$NettyEmbeddedServerInstance$Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractInitializableBeanDefinition.build(AbstractInitializableBeanDefinition.java:755)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2329)
	... 82 more
```

### Steps To Reproduce

1. Build sample app using `mn create-application` with JUnit, Gradle, Java 11/17
2. Add an empty second test method to example test class
3. Change class annotation to `@MicronautTest(rebuildContext = true)`
4. Attempt to run test class via gradle test

### Environment Information

- Tested on linux and macOS
- Java 11.0.7 and 17

### Example Application

_No response_

### Version

3.5.1

--- a/http-server-undertow/src/test/groovy/io/micronaut/servlet/undertow/EventsListener.java
+++ b/http-server-undertow/src/test/groovy/io/micronaut/servlet/undertow/EventsListener.java
@@ -1,0 +1,50 @@
+package io.micronaut.servlet.undertow;
+
+import io.micronaut.context.env.Environment;
+import io.micronaut.context.event.ApplicationEventListener;
+import io.micronaut.context.event.ShutdownEvent;
+import io.micronaut.context.event.StartupEvent;
+import io.micronaut.runtime.server.event.ServerShutdownEvent;
+import io.micronaut.runtime.server.event.ServerStartupEvent;
+import jakarta.inject.Singleton;
+
+@Singleton
+public class EventsListener implements ApplicationEventListener<Object>  {
+
+    public boolean serverStarted;
+    public boolean started;
+    public boolean serverShutdown;
+    public boolean shutdown;
+
+    public EventsListener(Environment environment) {
+        // This will fail if the event is sent when the bean context is shutdown
+    }
+
+    @Override
+    public void onApplicationEvent(Object event) {
+        if (event instanceof StartupEvent) {
+            if (started) {
+                throw new IllegalStateException();
+            }
+            started = true;
+        }
+        if (event instanceof ServerStartupEvent) {
+            if (!started || serverStarted) {
+                throw new IllegalStateException();
+            }
+            serverStarted = true;
+        }
+        if (event instanceof ShutdownEvent) {
+            if (!serverShutdown || shutdown) {
+                throw new IllegalStateException();
+            }
+            shutdown = true;
+        }
+        if (event instanceof ServerShutdownEvent) {
+            if (serverShutdown) {
+                throw new IllegalStateException();
+            }
+            serverShutdown = true;
+        }
+    }
+}

--- a/http-server-undertow/src/test/groovy/io/micronaut/servlet/undertow/UndertowStartStopSpec.groovy
+++ b/http-server-undertow/src/test/groovy/io/micronaut/servlet/undertow/UndertowStartStopSpec.groovy
@@ -1,0 +1,27 @@
+
+package io.micronaut.servlet.undertow
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.runtime.EmbeddedApplication
+import spock.lang.Specification
+
+class UndertowStartStopSpec extends Specification {
+
+    void "test the bean context is not used after shutdown"() {
+        when:
+            def ctx = ApplicationContext.builder().build()
+            ctx.start()
+            def eventsListener = ctx.getBean(EventsListener)
+            def embeddedApplication = ctx.getBean(EmbeddedApplication)
+            embeddedApplication.start()
+            embeddedApplication.stop()
+//            ctx.stop()
+        then:
+            !ctx.isRunning()
+            eventsListener.started
+            eventsListener.shutdown
+            eventsListener.serverStarted
+            eventsListener.serverShutdown
+    }
+
+}

--- a/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
+++ b/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
@@ -16,8 +16,6 @@
 package io.micronaut.servlet.engine.server;
 
 import io.micronaut.context.ApplicationContext;
-import io.micronaut.context.event.ApplicationEvent;
-import io.micronaut.context.event.ApplicationEventPublisher;
 import io.micronaut.discovery.ServiceInstance;
 import io.micronaut.discovery.event.ServiceReadyEvent;
 import io.micronaut.http.server.exceptions.HttpServerException;
@@ -37,7 +35,6 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
 
     private final ApplicationContext applicationContext;
     private final ApplicationConfiguration applicationConfiguration;
-    private final ApplicationEventPublisher<ApplicationEvent> applicationEventPublisher;
     private final T server;
 
     /**
@@ -54,7 +51,6 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
         this.applicationContext = applicationContext;
         this.applicationConfiguration = applicationConfiguration;
         this.server = server;
-        this.applicationEventPublisher = applicationContext.getEventPublisher(ApplicationEvent.class);
     }
 
     /**
@@ -81,7 +77,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
                 applicationContext.start();
             }
             startServer();
-            applicationEventPublisher.publishEvent(new ServerStartupEvent(this));
+            applicationContext.publishEvent(new ServerStartupEvent(this));
             applicationConfiguration.getName().ifPresent((name) -> {
                 ServiceInstance.Builder builder = ServiceInstance.builder(name, getURI());
                 ApplicationConfiguration.InstanceConfiguration instance = applicationConfiguration.getInstance();
@@ -89,7 +85,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
                 instance.getZone().ifPresent(builder::zone);
                 builder.metadata(instance.getMetadata());
                 instance.getId().ifPresent(builder::instanceId);
-                applicationEventPublisher.publishEvent(new ServiceReadyEvent(builder.build()));
+                applicationContext.publishEvent(new ServiceReadyEvent(builder.build()));
             });
         } catch (Exception e) {
             throw new HttpServerException(
@@ -104,7 +100,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
         if (isRunning()) {
             try {
                 stopServer();
-                applicationEventPublisher.publishEvent(new ServerShutdownEvent(this));
+                applicationContext.publishEvent(new ServerShutdownEvent(this));
                 if (applicationContext.isRunning()) {
                     applicationContext.stop();
                 }

--- a/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
+++ b/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
@@ -22,7 +22,6 @@ import io.micronaut.discovery.ServiceInstance;
 import io.micronaut.discovery.event.ServiceReadyEvent;
 import io.micronaut.http.server.exceptions.HttpServerException;
 import io.micronaut.runtime.ApplicationConfiguration;
-import io.micronaut.runtime.event.ApplicationStartupEvent;
 import io.micronaut.runtime.server.EmbeddedServer;
 import io.micronaut.runtime.server.event.ServerShutdownEvent;
 import io.micronaut.runtime.server.event.ServerStartupEvent;

--- a/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
+++ b/servlet-engine/src/main/java/io/micronaut/servlet/engine/server/AbstractServletServer.java
@@ -16,10 +16,13 @@
 package io.micronaut.servlet.engine.server;
 
 import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.event.ApplicationEvent;
+import io.micronaut.context.event.ApplicationEventPublisher;
 import io.micronaut.discovery.ServiceInstance;
 import io.micronaut.discovery.event.ServiceReadyEvent;
 import io.micronaut.http.server.exceptions.HttpServerException;
 import io.micronaut.runtime.ApplicationConfiguration;
+import io.micronaut.runtime.event.ApplicationStartupEvent;
 import io.micronaut.runtime.server.EmbeddedServer;
 import io.micronaut.runtime.server.event.ServerShutdownEvent;
 import io.micronaut.runtime.server.event.ServerStartupEvent;
@@ -35,6 +38,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
 
     private final ApplicationContext applicationContext;
     private final ApplicationConfiguration applicationConfiguration;
+    private final ApplicationEventPublisher<ApplicationEvent> applicationEventPublisher;
     private final T server;
 
     /**
@@ -51,6 +55,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
         this.applicationContext = applicationContext;
         this.applicationConfiguration = applicationConfiguration;
         this.server = server;
+        this.applicationEventPublisher = applicationContext.getEventPublisher(ApplicationEvent.class);
     }
 
     /**
@@ -77,7 +82,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
                 applicationContext.start();
             }
             startServer();
-            applicationContext.publishEvent(new ServerStartupEvent(this));
+            applicationEventPublisher.publishEvent(new ServerStartupEvent(this));
             applicationConfiguration.getName().ifPresent((name) -> {
                 ServiceInstance.Builder builder = ServiceInstance.builder(name, getURI());
                 ApplicationConfiguration.InstanceConfiguration instance = applicationConfiguration.getInstance();
@@ -85,7 +90,7 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
                 instance.getZone().ifPresent(builder::zone);
                 builder.metadata(instance.getMetadata());
                 instance.getId().ifPresent(builder::instanceId);
-                applicationContext.publishEvent(new ServiceReadyEvent(builder.build()));
+                applicationEventPublisher.publishEvent(new ServiceReadyEvent(builder.build()));
             });
         } catch (Exception e) {
             throw new HttpServerException(
@@ -100,10 +105,10 @@ public abstract class AbstractServletServer<T> implements EmbeddedServer {
         if (isRunning()) {
             try {
                 stopServer();
+                applicationEventPublisher.publishEvent(new ServerShutdownEvent(this));
                 if (applicationContext.isRunning()) {
                     applicationContext.stop();
                 }
-                applicationContext.publishEvent(new ServerShutdownEvent(this));
             } catch (Exception e) {
                 throw new HttpServerException(
                         "Error stopping HTTP server: " + e.getMessage(), e

